InformationScripting/src/graph/Graph.cpp

```diff
@@ -0,0 +1,191 @@
+			existingEdge->incrementCount();
+			return existingEdge;
+		}
+	}
+	auto edge = new InformationEdge(from, to, name);
+	edges_.push_back(edge);
+	return edge;
+}
+
+InformationEdge* Graph::addEdge(InformationNode* a, InformationNode* b, const QString& name)
+{
+	// We only allow existing nodes:
+	Q_ASSERT(a == findNode(a));
+	Q_ASSERT(b == findNode(b));
+	for (auto edge : edges_)
+	{
+		if ((edge->from() == a || edge->from() == b) && (edge->to() == b || edge->to() == a) && edge->name() == name)
+		{
+			auto existingEdge = edge;
+			Q_ASSERT(!existingEdge->isDirected());
+			existingEdge->incrementCount();
+			return existingEdge;
+		}
+	}
+	auto edge = new InformationEdge(a, b, name, InformationEdge::Orientation::Undirected);
+	edges_.push_back(edge);
+	return edge;
+}
+
+void Graph::remove(InformationNode* node)
+{
+	Q_ASSERT(node);
+	auto hash = hashValueOf(node);
+	auto nodeIt = nodes_.find(hash);
+	// If the node isn't in the graph we don't have to do anything:
+	if (nodeIt == nodes_.end()) return;
+
+	// It could be that node points to a different node with the same hash.
+	// TODO in this case we might also want to merge the infos?
+	auto ownedNode = node;
+	if (ownedNode != *nodeIt) ownedNode = *nodeIt;
+
+	for (auto edgeIt = edges_.begin(); edgeIt != edges_.end();)
+	{
+		auto edge = *edgeIt;
+		if (edge->from() == ownedNode || edge->to() == ownedNode)
+		{
+			SAFE_DELETE(edge);
+			edgeIt = edges_.erase(edgeIt);
+		}
+		else
+		{
+			++edgeIt;
+		}
+	}
+
+	SAFE_DELETE(ownedNode);
+	nodes_.erase(nodeIt);
+}
+
+void Graph::remove(QList<InformationNode*> nodes)
+{
+	for (auto node : nodes) remove(node);
+}
+
+QList<InformationNode*> Graph::nodes() const
+{
+	return nodes_.values();
+}
+
+QList<InformationNode*> Graph::nodesForWhich(Graph::NodeCondition holds) const
+{
+	Q_ASSERT(holds);
+
+	QList<InformationNode*> result;
+	for (auto node : nodes_) if (holds(node)) result.push_back(node);
+	return result;
+}
+
+QList<InformationEdge*> Graph::edgesFowWhich(Graph::EdgeCondition holds) const
+{
+	Q_ASSERT(holds);
+
+	QList<InformationEdge*> result;
+	for (auto edge : edges_) if (holds(edge)) result.push_back(edge);
+	return result;
+}
+
+std::size_t Graph::hashValueOf(const InformationNode* n) const
+{
+	for (auto hashFunction : nodeHashFunctions_)
+	{
+		auto result = hashFunction(n);
+		if (result.second) return result.first;
+	}
+	// There should be at least one function registered for a certain node
+	Q_ASSERT(false);
+}
+
+InformationNode*Graph::findNode(const InformationNode* n) const
+{
+	auto hash = hashValueOf(n);
+	auto it = nodes_.find(hash);
+	if (it != nodes_.end()) return it.value();
+	return nullptr;
+}
+
+void Graph::mergeNodes(InformationNode* into, InformationNode* from)
+{
+	for (auto namedProperty : *from)
+	{
+		QString name = namedProperty.first;
+		if (!into->contains(name))
+			into->insert(name, namedProperty.second);
+		else
+			Q_ASSERT((*from)[name] == namedProperty.second);
```

Comment:
This is OK for now, but down the road, we might need to make the `into` or `from` node dominant, which would mean that in case of conflict we just pick its value.